### PR TITLE
rgw: use unique_ptr for flat_map emplace in BucketTrimWatche

### PR DESCRIPTION
**Pull request by vedanshbhartia**
## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/rgw/driver/rados/rgw_trim_bilog.cc [hunk 1]
```diff
@@ -253,8 +253,20 @@ class BucketTrimWatcher : public librados::WatchCtx2 {
...
                     TrimCounters::Server *counters)
     : store(store), obj(obj) {
-    handlers.emplace(NotifyTrimCounters, new TrimCounters::Handler(counters));
-    handlers.emplace(NotifyTrimComplete, new TrimComplete::Handler(counters));
+    std::unique_ptr<TrimCounters::Handler> counter_handler(
+        new TrimCounters::Handler(counters));
+    std::unique_ptr<TrimComplete::Handler> complete_handler(
+        new TrimComplete::Handler(counters));
+
+    auto result = handlers.emplace(NotifyTrimCounters, std::move(counter_handler));
```
Comment (cbodley) [inline src/rgw/driver/rados/rgw_trim_bilog.cc:261]: can we just replace the `new TrimCounters::Handler(counters)` arguments with `std::make_unique<TrimCounters::Handler>(counters)`? that should free the memory whether or not the `emplace()` succeeds

**Comment by mattbenjamin** [comment]
@vedanshbhartia just out of curiosity, what motivated you to focus on this change.  If I understand your commit msg and the tracker remarks correctly, nothing is leaked here because of the behavior of flat_map.  It would have leaked if the code used a container it does not use.  Are you looking at other areas in which this idiom has been found to leak?

**Review by cbodley — APPROVED** [review]
looks great. thanks for the detailed commit message

**Comment by cbodley** [comment]
passed qa in https://pulpito.ceph.com/cbodley-2023-03-10_19:16:19-rgw-wip-cbodley-testing-distro-default-smithi/

**Pull request merged**